test/tls-provider.c

```diff
@@ -748,89 +1274,2261 @@ static const OSSL_DISPATCH xor_keymgmt_functions[] = {
+                    der, PrivateKeyInfo),
+ENCODER_w_structure(XORSIGALG_HASH_NAME, xorhmacsha2sig,
+                    pem, PrivateKeyInfo),
+ENCODER_w_structure(XORSIGALG_HASH_NAME, xorhmacsha2sig,
+                    der, EncryptedPrivateKeyInfo),
+ENCODER_w_structure(XORSIGALG_HASH_NAME, xorhmacsha2sig,
+                    pem, EncryptedPrivateKeyInfo),
+ENCODER_w_structure(XORSIGALG_HASH_NAME, xorhmacsha2sig,
+                    der, SubjectPublicKeyInfo),
+ENCODER_w_structure(XORSIGALG_HASH_NAME, xorhmacsha2sig,
+                    pem, SubjectPublicKeyInfo),
+#undef ENCODER_PROVIDER
+    { NULL, NULL, NULL }
+};
+
+struct der2key_ctx_st;           /* Forward declaration */
+typedef int check_key_fn(void *, struct der2key_ctx_st *ctx);
+typedef void adjust_key_fn(void *, struct der2key_ctx_st *ctx);
+typedef void free_key_fn(void *);
+typedef void *d2i_PKCS8_fn(void **, const unsigned char **, long,
+                           struct der2key_ctx_st *);
+struct keytype_desc_st {
+    const char *keytype_name;
+    const OSSL_DISPATCH *fns; /* Keymgmt (to pilfer functions from) */
+
+    /* The input structure name */
+    const char *structure_name;
+
+    /*
+     * The EVP_PKEY_xxx type macro.  Should be zero for type specific
+     * structures, non-zero when the outermost structure is PKCS#8 or
+     * SubjectPublicKeyInfo.  This determines which of the function
+     * pointers below will be used.
+     */
+    int evp_type;
+
+    /* The selection mask for OSSL_FUNC_decoder_does_selection() */
+    int selection_mask;
+
+    /* For type specific decoders, we use the corresponding d2i */
+    d2i_of_void *d2i_private_key; /* From type-specific DER */
+    d2i_of_void *d2i_public_key;  /* From type-specific DER */
+    d2i_of_void *d2i_key_params;  /* From type-specific DER */
+    d2i_PKCS8_fn *d2i_PKCS8;      /* Wrapped in a PrivateKeyInfo */
+    d2i_of_void *d2i_PUBKEY;      /* Wrapped in a SubjectPublicKeyInfo */
+
+    /*
+     * For any key, we may need to check that the key meets expectations.
+     * This is useful when the same functions can decode several variants
+     * of a key.
+     */
+    check_key_fn *check_key;
+
+    /*
+     * For any key, we may need to make provider specific adjustments, such
+     * as ensure the key carries the correct library context.
+     */
+    adjust_key_fn *adjust_key;
+    /* {type}_free() */
+    free_key_fn *free_key;
+};
+
+/*
+ * Start blatant code steal. Alternative: Open up d2i_X509_PUBKEY_INTERNAL
+ * as per https://github.com/openssl/openssl/issues/16697 (TBD)
+ * Code from from openssl/crypto/x509/x_pubkey.c as
+ * ossl_d2i_X509_PUBKEY_INTERNAL is presently not public
+ */
+struct X509_pubkey_st {
+    X509_ALGOR *algor;
+    ASN1_BIT_STRING *public_key;
+
+    EVP_PKEY *pkey;
+
+    /* extra data for the callback, used by d2i_PUBKEY_ex */
+    OSSL_LIB_CTX *libctx;
+    char *propq;
+
+    /* Flag to force legacy keys */
+    unsigned int flag_force_legacy : 1;
+};
```

Comment:
This looks suspect to me and I'm not too sure why its necessary???

@levitte may have some insight here

Reply:
Good point, it's not.   I guess that's part of the steal (this is, after all, a verbatim copy from `crypto/x509/x_pubkey.c`)

Reply:
It was the only way I found for re-creating the internal key structure from the X509 structure. This may very well be due to my lack of understanding of (X509) APIs, so if there's better alternatives, I'm all ears, @mattcaswell @levitte

Reply:
[ahem] I haven't really made a deep dive into the added test code, so can't answer for the moment.  Will do when I have a moment, probably this afternoon.

Reply:
@levitte your input here would still be helpful...

Reply:
I never got to looking at this in depth, will try to do so this week

Reply:
So, what @mattcaswell wants my opinion of is that flag, specifically: `flag_force_legacy`

It does look unnecessary, as it's not touched anywhere, and this code is all about providers anyway, so why have that flag at all?  The code in `crypto/x509/x_pubkey.c` is different in this regard, it *has* to be able to deal with legacy keys as well.

Reply:
Correct. Removed in https://github.com/openssl/openssl/pull/19312/commits/280f9d4bcea83a45c4ea4024a9a5aada6d177eaa.